gpu_ocean/SWESimulators/gpu_kernels/CDKLM16_kernel.cu

```diff
@@ -358,18 +578,45 @@ __global__ void cdklm_swe_2D(
         for (int i=tx; i<block_width+4; i+=blockDim.x) {
             const int k = clamp(bx+i, 0, nx_+3);
 
-            //const float h = R[0][j][i] + Hm[j][i]; // h = eta + H
-            const float h = R[0][j][i] + Hm_row[k];
-            R[1][j][i] /= h;
-            R[2][j][i] /= h;
+            // h = eta + H
+            const float local_Hm = Hm_row[k];
+            //const float local_Hm =  0.25f*(Hi[ty+1][tx+1]+Hi[ty+2][tx+1]+Hi[ty+1][tx+2]+Hi[ty+2][tx+2]);
+            const float h = R[0][j][i] + local_Hm;
+            
+           
+            
+            // Check if the cell is almost dry
+            if (h < KPSIMULATOR_FLUX_SLOPE_EPS) {
+                
+                if (h <= KPSIMULATOR_DEPTH_CUTOFF) {
+                    //R[0][j][i] = -local_Hm;
+                    R[1][j][i] = 0.0f;
+                    R[2][j][i] = 0.0f;
+                }
+                else {                
+                    // Desingularizing u and v
+                    float h4 = h*h; h4 *= h4;
+                    R[1][j][i] = SQRT_OF_TWO*h*R[1][j][i]/sqrt(h4 + fmaxf(h4, KPSIMULATOR_FLUX_SLOPE_EPS_4));
```

Comment:
Maybe we could use the new desingularization here instead, and add it to the paper?